### DISCUSSION TITLE
how to deal with dma and buffer alignment

### DISCUSSION THREAD

**Discussion by cleverca22** [hathach/tinyusb#2189] [Q&A]
```c
typedef struct
{
  uint8_t itf_num;
  uint8_t ep_in;
  uint8_t port_count;
  uint8_t status_change; // data from status change interrupt endpoint

  CFG_TUH_MEM_ALIGN hub_port_status_response_t port_status;
  CFG_TUH_MEM_ALIGN hub_status_response_t hub_status;
} hub_interface_t;
```
and some debug logging:
```
 20.461786 [DWC2:hcd_edpt_xfer:393]:    HOST0  <-   5.81 0xc4015c3b/1 opep:0xc4114784 type:3 pid:0
 24.332040 [DWC2:hcd_edpt_xfer:393]:    HOST0  <-   5.c1 0xc4015c3b/1 opep:0xc4114784 type:3 pid:0
```
the log, says that tinyusb requested a 1 byte IN from endpoint 81 on device 5 (the endpoint c1!)
the 1 byte should be written to 0xc4015c3b (the `status_change` field)

but because the dma requires 32bit aligned access, it rounded down, and wrote to `itf_num` as well, corrupting the EP# and causing the next transfer to be on endpoint c1

the question then, is how to best deal with this?

the simplest option i can see, is to check if the length is a multiple of 32bits, and the buffer is a multiple of 32bits
if not, dma to a safe space, then `memcpy()` over, harming performance

a better option, is for tinyusb to have a special api, to make all buffers safe, or to tell the hcd when a buffer is safe

**Comment by cleverca22** [comment]
on a second look, what if i just slap that `CFG_TUH_MEM_ALIGN` on `status_change`....
that might fix this issue, but no telling where it will crop up again, and the field after any buffer must also be aligned, to create a safe space

that does at least fix the hub corruption

**Comment by hathach** [comment]
latest code already has status_changge with mem align, please pull update and try again

**Comment by cleverca22** [comment]
same problem exists in the msc class
```
  uint8_t ep_out;
  
  CFG_TUH_MEM_ALIGN uint8_t max_lun;

  CFG_TUH_MEM_ALIGN volatile bool configured; // Receive SET_CONFIGURE
  volatile bool mounted;    // Enumeration is complete
```
`max_lun` must be aligned for the transfer to work
`configured` must be aligned, or a 32bit write to `max_lun` will trash it and more

**Reply by hathach** [reply]
thank you for spotting it out, fixed this in latest push #2211 . only max_lun is required to be aligned, but I have use the msch buffer instead.
